### ISSUE TITLE
Something wrong with new model generating 128 embeddings?

### ISSUE THREAD

**Issue by dleonsanchez** [davidsandberg/facenet#189]
Hi @davidsandberg,

Thanks for sharing the new model that generates 128 embeddings.

I am already testing it and there is something that is puzzling me: is it possible that somehow this new model only provides meaningful embeddings for images that are forward-passed in the same sess.run execution? If I run a batch process calling sess.run multiple times (say with one face at a time), the embeddings are so close to each other that the euclidean distance between different faces is 0.0000006 or so, and the accuracy is simply not good. If I pass those faces in the same sess.run in the images_placeholder input array (as in the example in compare.py), then the embeddings make sense and the distances are much greater.

I guess I am doing something wrong, but I just don't seem to see what it is exactly. You can reproduce what I describe if you do the following in compare.py:

Instead of:
```
feed_dict = { images_placeholder: images, phase_train_placeholder:False }
emb = sess.run(embeddings, feed_dict=feed_dict)

```

This:
```
emb = np.zeros((nrof_images, 128))
for i in range(nrof_images):
        npimages = np.stack([images[i]])
        feed_dict = { images_placeholder: npimages, phase_train_placeholder:False }
        emb[i, :]  = sess.run(embeddings, feed_dict=feed_dict)

```
The result is that the distance matrix is 0.0000 in all cases.

Any ideas?

**Comment by davidsandberg** [comment]
Hi @dleonsanchez,
Thanks for spotting this!!!
I believe this is caused by the batch norm for the bottleneck layer always being in training mode.
it should be possible to fix by setting the parameter `is_training`, something like this:
```
        batch_norm_params = {
            # Decay for the moving averages.
            'decay': 0.995,
            # epsilon to prevent 0s in variance.
            'epsilon': 0.001,
            # force in-place updates of mean and variance estimates
            'updates_collections': None,
            # Moving averages ends up in the trainable variables collection
            'variables_collections': [ tf.GraphKeys.TRAINABLE_VARIABLES ],
            # Only update statistics during training mode
            'is_training': phase_train_placeholder
        }
```
I haven't tested this yet though.

**Comment by ugtony** [comment]
I've tested @davidsandberg's method, and it solved the problem. 
Thanks for the hint.

**Comment by RakshakTalwar** [comment]
@ugtony can you please elaborate on how you fixed it?
I seem to be spinning my wheels here

Thank you very much

**Comment by ugtony** [comment]
@RakshakTalwar,
First, I modified facenet_train_classifier.py as what davidsandberg's had suggested. 
And then I used that script to save the model after the preload_model was restored.
The script for test time was kept unchanged.

I think it is also possible to directly manipulate the graph after import_meta_graph(), but it's more difficult for me to do that.

**Comment by AKSHAYUBHAT** [comment]
@ugtony Thank you so much I literally spent last 4 hours wondering why none of the results made any sense, there were surely bugs in my code like cropping instead of scaling, not whitening in same manner, but eventually it wouldn't work in spite of having a complete blank image. As soon as I changed model files with the link you provided it worked beautifully!

**Comment by ugtony** [comment]
@AKSHAYUBHAT 
You're welcome. 
I was also checking cropping, resizing, and whitening as you did. And then this opened issue came to my head and solved my problem. Thank dleonsanchez 's question and davidsandberg 's solution that saved my time.

**Comment by davidsandberg** [comment]
Thanks @ugtony etc!
I committed this fix and updated the meta files for the pretrained models. I hope it works...
